### DISCUSSION TITLE
Keyboard shortcut for reconnecting as a way to fix locking RDP session

### DISCUSSION THREAD

**Discussion by rafi-d** [1Remote/1Remote#659] [General]
Hi,
Is there a list of key shortcuts available for 1remote manager? 
Specifically, I like the shortcut for "**reconnect**" menu item (for the selected PC[s]), for my frequent use case:
The remote is locking the session into the login screen, and re-connect is the easiest way to re-login... 

...and If possible - for "connect" too?

There was an old request here for user-definable hot keys, but the above is the minimum for me :) ...
There was also a request of mine for preventing lock timeout (  #629 ) which this can be a semi work-around to ... BTW, emulating key/mouse input/juggle  option might work for preventing timeouts if you can generate such inputs

**Comment by majkinetor** [comment]
There is also #405.

**Comment by majkinetor** [comment]
> BTW, emulating key/mouse input/juggle option might work for preventing timeouts if you can generate such inputs

Good idea, it never occurred to me. Emulating keys is IMO too intrusive, as that can influence the remote. Moving a mouse, though, slightly, then moving it in the opposite direction could work.

**Comment by rafi-d** [comment]
> Emulating keys is IMO too intrusive, as that can influence the remote. Moving a mouse, though, slightly,

Right, possibly for a "timeout" duration... It might work, but needs testing, since it is not clear if a mouse move by itself, without any action (like click+scroll)  will count as an "activity" ... 
A key pressed is always an activity, so maybe some "harmless" key-code can be found .. 

> There is also #405.

Yeah, I was referring to this one...

**Comment by majkinetor** [comment]
I doubt the harmless key exists - you could have a keyloger for example. 

You can check out if mouse works by using [Autohotkey](https://www.autohotkey.com)  on the remote host:

Save this as mouse.ahk and doubleclick it to run it, after you install AHK (portable version can be used too).  You can see the script running in the tray.

```ahk
loop {
    sleep 60000
    MouseMove 1, 0, 50, R
    MouseMove -1, 0, 50, R
}
```
This will move mouse 1 pixel left then right every 60s, making it keep its previous position. Its good enough for PoC.

If this works, I guess 1remote can simulate this as background task based on configuration option, but @VShawn will know more about it.

**Comment by majkinetor** [comment]
BTW, I think reconnect hotkey might not be possible - once remote grabs input, its AFAIK not possible to execute some hotkeys on the local host, since after locking you out, remote will still keep keyboard focus on the login UI, and those hotkeys would finish on remote (see #40)

**Comment by rafi-d** [comment]
Again, it needs testing. 
I imagine - that like you defined "control-F" to work for search on the admin GUI when in focus (not inside an RDP session), you might be able to allocate , say, control-R to work on the currently selected host (s) as 'reconnect' or connect (e.e. similar to activating the menu-> reconnect)

Trying this script , see if it works:

#Requires AutoHotkey >=1.0.0
loop {
    sleep 60000
    MouseMove  10, 0, 50, "R"
   MouseMove  -10, 0, 50, "R"
}

But I think it might only work if the *host* moves the mouse, and not the client itself...

**Comment by rafi-d** [comment]
^ First test seems to go well: running this script *locally* on the host *without* RDP (with an active open screen...) . No screen-lock up ... 
Later I'll leave it running with RDP.

Edit: Same result. Connection stayed open.  
So, your test is - try use it from the host side (remote mouse) ...

**Reply by majkinetor** [reply]
OK, so moving the mouse prevents locking, you can use that script to prevent it for now.

But I think that 1Remote can also know the current mouse position in the remote window and do the same as AHK script. Not entirely sure though, @VShawn will tell.

**Reply by VShawn** [reply]
I saw an app that could send mouse movement commands to multiple RDP desktops in batch, but I don't know how it was done, and I can't find that app anymore.

**Reply by majkinetor** [reply]
I guess it is done by sending the [wm-mousemove](https://learn.microsoft.com/en-us/windows/win32/inputdev/wm-mousemove) to adequate window. Of course, particular window doesn't have to be visible for this.

This can be done with AHK script which does that, as shown by rafi-d [here](https://github.com/1Remote/1Remote/discussions/659#discussioncomment-9313060)

**Reply by majkinetor** [reply]
Now, not sure if it works in background/inactive window, just like that. It might require [Windows hook](https://www.autohotkey.com/boards/viewtopic.php?p=493728#p493728)

**Reply by majkinetor** [reply]
It might be problematic to move a mouse in a background window. Sending keys to background window, on the other hand, seems to work without problems.  So it may be easier to achieve pressing CTRL or ALT that should do nothing...

**Comment by rafi-d** [comment]
I did try this also indirectly - I did NOT run any script on the remote, opened 1 remote, and ran the same script on the local PC *WHILE* it was open and the mouse was in that Window area. It worked too.So yes, if you can do the same programmatically - it can be great. Just give an option to define both the loop time (5 min might be enough for most cases...) and possibly the total duration (timeout, or - "forever") . Not everyone wants it running forever....

The open question about a hot-key for menu items (reconnect / connect ) is still interesting/useful  too I think...

**Comment by rafi-d** [comment]
One more related problem: Once the remote is locked, the RDP display in full screen will NOT show the locked screen but the screen before. This is VERY confusing and unfriendly :( 
Is there any setting I should change to avoid this?

**Reply by majkinetor** [reply]
I never got that, so I couldn't say. It always shows lock screen to me.

**Comment by rafi-d** [comment]
Just login to whatever and wait (inactive) with full screen, and minimize it. After the timeout, it will be unresponsive (aka locked but not visibly locked...) , unless you re-size it

**Comment by VShawn** [comment]
I have never set a timeout for my server, so I am curious why you don't set a longer timeout for RDP?

- https://answers.microsoft.com/en-us/windowserver/forum/all/remote-desktop-how-to-increase-lock-timeout/fc0f76a4-6a48-41f1-95d8-fbbc4e6a2ae9
- https://superuser.com/questions/86244/timeout-settings-for-remote-desktop-sessions-to-lock

**Reply by majkinetor** [reply]
You might not have a control of such server options

**Comment by rafi-d** [comment]
> I have never set a timeout for my server, so I am curious why you don't set a longer timeout for RDP?

Well, it is not always *your* remote "server" or work-station in my case. My use-case is two remote work-PCs, where IT of those companies do not let you change any of the security related settings (Group Policy). One of these - is lock timeout. You simply *cannot* change it since it is considered  security related. As I work at home, both machines a local, and accessible via RDP. Lock screen is an annoyance... The only way to W/O is (beside a convinient way to re-connect) is to W/A the predefined lock timeout (in fact - cancel it or make it several hours). 

So, this Autohotkey app - is definitely a W/A for me, but cannot be used on the RDP controller/PC... 

>I saw an app that could send mouse movement commands to multiple RDP desktops in batch, 

I am not sure why you need an external app, isn't it possible from your own RDP "manager" app?

**Comment by rafi-d** [comment]
>  So it may be easier to achieve pressing CTRL or ALT that should do nothing...

Or maybe move-up->down arrows , like with the mouse... 
Anyway, it is expected of the main application to keep track of both tabs OR/AND windows, so to be able to send/do anything to them (using the newly defined per-RDP-connection/window settings...

**Reply by majkinetor** [reply]
move up then down is not OK.

Lets say you are in the browser at the beginning of the page. Move up does nothing, but move down scrolls 1 step down, so you changed the view. Subsequent up/down will keep the new view though, but this illustrates the point.

**Reply by rafi-d** [reply]
Same issue with mouse move when it is located at the very right or left screen edge and the first move is outside of it... :)

**Reply by majkinetor** [reply]
Indeed. But it won't change any underlying app in any noticable way :)

**Reply by rafi-d** [reply]
I agree. Now let see it being implemented by the 1Remote app...
